[controller & command] training media cache

fenrir-z opened the pull request:
# changes
* mir train: add --asset-cache-dir
* controller invoker for training: add media cache
* added hash id sub directory in data exporter in command training

Commit:
WIP: training add cache

Commit:
Merge branch 'dev' into zzw/dev-cmd-train-cache

Commit:
wip

Commit:
Merge branch 'dev' into zzw/dev-cmd-train-cache

Commit:
remove unused func

Commit:
Merge branch 'dev' into zzw/dev-cmd-train-cache

Commit:
data exporter: also export annotations into sub id dir;
mir train: some errors by return code changed to raise errors

Commit:
remove unused logging

Commit:
training invoker added assets cache

Commit:
remove unused imports

Commit:
Merge branch 'dev' into zzw/dev-cmd-train-cache

Commit:
fix mypy and flake8 errors

fenrir-z committed:
Merge branch 'dev' into zzw/dev-cmd-train-cache

Commit:
Merge branch 'dev' into zzw/dev-cmd-train-cache

Commit:
data exporter generated index file also contains annotation file path

fenrir-z committed:
Merge branch 'dev' into zzw/dev-cmd-train-cache

fenrir-z committed:
Merge branch 'dev' into zzw/dev-cmd-train-cache

Commit:
data exporter: remove unused todos

Commit:
Merge branch 'zzw/dev-cmd-train-cache' of github.com:fenrir-z/ymir into zzw/dev-cmd-train-cache

Commit:
phase logger: fix todo

phoenix-xhuang commented on ymir/backend/src/ymir_controller/controller/invoker/invoker_task_training.py:
```diff
@@ -110,7 +113,7 @@ def training_cmd(
             '--model-location', models_upload_location, '--media-location', media_location, '-w', work_dir,
             '--src-revs', f"{in_dataset_id}@{his_task_id}", '--config-file', config_file, '--executor', training_image,
-            '--executor-instance', executor_instance, '--tensorboard', tensorboard
+            '--executor-instance', executor_instance, '--tensorboard', tensorboard, '--cache', asset_cache_dir,
```
s/cache/cache_dir

fenrir-z replied:
mining 任务的参数名称也是 cache
而且其他与目录有类的参数，比如 tensorboard，后面也没加 dir

phoenix-xhuang replied:
这里cache的名称缺乏可读性，叫asset_cache_dir更合适。tensorboard要改成tensorboard_dir也行，只是问题没cache这样明显

phoenix-xhuang replied:
最好是把mining的也一起改了

fenrir-z replied:
updated

fenrir-z committed:
Merge branch 'dev' into zzw/dev-cmd-train-cache

fenrir-z committed:
Merge branch 'dev' into zzw/dev-cmd-train-cache

phoenix-xhuang commented on ymir/command/mir/commands/training.py:
```diff
@@ -172,6 +173,7 @@ def run(self) -> int:
     @staticmethod
     @command_run_in_out
     def run_with_args(work_dir: str,
+                      media_cache: Optional[str],
```
media_cache和asset_chache_dir是啥关系

fenrir-z replied:
命令行参数是 --asset-cache-dir，对应的 args 成员是 media_cache
我统一一下

fenrir-z replied:
done

phoenix-xhuang commented on ymir/command/mir/commands/training.py:
```diff
@@ -362,6 +368,8 @@ def run_with_args(work_dir: str,
         path_binds = []
         path_binds.append(f"-v{work_dir_in}:/in")
+        if media_cache:
+            path_binds.append(f"-v{media_cache}:/in/cache")
```
能否把cache做成对镜像是透明的，不需要感知到

fenrir-z replied:
照现在这种映射方式，有点麻烦。

我们能不能这样映射：

asset_cache_dir or work_dir/in/assets -> /in/assets (rx)

work_dir/in/labels -> /in/labels (rwx)

/in/train-index.tsv (rwx)

/in/val-index.tsv (rwx)

这样可以保证一定的透明性

phoenix-xhuang replied:
这样的映射方式的确更好，正好要准备训练镜像框架，可以按这个思路一起调整下

fenrir-z replied:
我顺便把 mining & infer 那边的映射方式一起调整成类似的。
核心的思想是：
1. 所有的 assets 和 labels 分开存放，其中 assets 目录是只读的，而 labels 目录是可以写入的
2. index.tsv 文件根据文件名区分用途，训练集为 train-index.tsv，验证集为 val-index.tsv，测试集为 test-index.tsv，mining 或 infer 的候选集为 candidate-index.tsv，都在 /in 目录下

phoenix-xhuang replied:
其它没问题，“ labels 目录是可以写入的”是啥意思？镜像可以写入还是YMIR可以写入？

fenrir-z replied:
镜像是可以写入的

Commit:
mir train:
    --cache -> --asset-cache-dir
    --tensorboard -> --tensorboard-dir
mir mining:
    --cache -> assest-cache-dir
invoker task training & invoker task mining also changed

Commit:
mir train & mir mining: tiny changed

Commit:
labels.py: bug fix

Commit:
mir train: make cache dir readonly

Commit:
Merge branch 'dev' into zzw/dev-cmd-train-cache

Commit:
mir train: update directory and file binding

Commit:
mir mining & infer: update dir bindings; test cases passed;
mir training: removed unused imports

Commit:
Merge branch 'dev' into zzw/dev-cmd-train-cache

fenrir-z committed:
Merge branch 'dev' into zzw/dev-cmd-train-cache

fenrir-z committed:
Merge branch 'dev' into zzw/dev-cmd-train-cache

Commit:
update container doc

Commit:
disabled changes to labels.py

fenrir-z committed:
Merge branch 'dev' into zzw/dev-cmd-train-cache

Commit:
Merge branch 'dev' into zzw/dev-cmd-train-cache

phoenix-xhuang committed:
Merge branch 'dev' into zzw/dev-cmd-train-cache

Commit:
Merge branch 'dev' into zzw/dev-cmd-train-cache

Commit:
Merge branch 'zzw/dev-cmd-train-cache' of github.com:fenrir-z/ymir into zzw/dev-cmd-train-cache

fenrir-z committed:
Merge branch 'dev' into zzw/dev-cmd-train-cache

phoenix-xhuang commented:
方向上的问题，上次讨论过，不要让镜像感知到cache的存在，对吧？原因在于：cache是YMIR自已的组织模式，镜像不需要感知到。YMIR-CMD在挂载和生成index的时候，做出是否要cache的判断，但都是统一放在/in/assets目录下

phoenix-xhuang commented:
补充下，是assets放到/in/assets目录下，index文件现在放的地方没问题

fenrir-z committed:
Merge branch 'dev' into zzw/dev-cmd-train-cache

fenrir-z commented:
> 补充下，是assets放到/in/assets目录下，index文件现在放的地方没问题

对，现在就是这样的行为。
所有 assets，不管它是在 cache 下面，还是真的在 <work_dir>/in/assets 下面，最终挂载的位置都是 /in/assets，这一点在 bind path 的设置里面就可以看出来。
这个行为，在 training 和 mining 里面都是一样的

phoenix-xhuang commented:
看到了，executor里面写的是以前的行为

fenrir-z commented:
> 看到了，executor里面写的是以前的行为

是不是有啥地方没修改完全？

fenrir-z committed:
Merge branch 'dev' into zzw/dev-cmd-train-cache

fenrir-z commented:
> 
executor 里面我应该已经删掉了所有关于 /in/cache 的描述

phoenix-xhuang commented on docs/ymir-cmd-container.md:
```diff
@@ -191,10 +191,12 @@ task_0    1622552975    1    done
-| /in/config.yaml | 必要，配置项所在的文件，YAML格式，具体配置项由调用双方商定，注1中列出了几个固定的键 |
+| /in/train-index.tsv | 必要，训练集图像及标注的索引文件，每一行都是图片文件路径和标注路径<br>* 图片文件为jpg，png等常用格式（但不保证一定有扩展名）；<br>* 标注文件可以是csv，txt，xml，json格式，拥有和图片相同的主文件名。 |
+| /in/val-index.tsv | 必要，验证集图像及标注的索引文件，格式同/in/train-index.tsv |
+| /in/test-index.tsv | 非必要，测试集图像及标注的索引文件，格式同/in/train-index.tsv |
```
看上去是不需要这个/in/test-index.tsv了，我们也没有接口输出这个

fenrir-z replied:
removed

phoenix-xhuang replied:
回退的时候是把这个也恢复了？

fenrir-z replied:
这里忘记调整了

phoenix-xhuang commented on docs/ymir-cmd-container.md:
```diff
@@ -191,10 +191,12 @@ task_0    1622552975    1    done
-| /in/val | 必要，验证用图像资源及标注对所在的目录，格式同/in/train |
-| /in/test | 非必要，测试用图像资源及标注对所在的目录，格式同/in/train |
-| /in/config.yaml | 必要，配置项所在的文件，YAML格式，具体配置项由调用双方商定，注1中列出了几个固定的键 |
+| /in/train-index.tsv | 必要，训练集图像及标注的索引文件，每一行都是图片文件路径和标注路径<br>* 图片文件为jpg，png等常用格式（但不保证一定有扩展名）；<br>* 标注文件可以是csv，txt，xml，json格式，拥有和图片相同的主文件名。 |
```
是否要考虑把/in目录做成只读

fenrir-z replied:
done

fenrir-z replied:
想了一下，回退了。
之所以把 assets 目录做成 readonly 的，只是因为它有可能是 cache 目录，而 cache 目录是不能更改的。
其他目录都没有这种强制性规定。

phoenix-xhuang commented on ymir/backend/src/ymir_controller/controller/invoker/invoker_task_mining.py:
```diff
@@ -112,8 +112,8 @@ def mining_cmd(
-            '--src-revs', f"{in_dataset_id}@{his_task_id}", '--cache', asset_cache_dir, '--config-file', config_file,
-            '--executor', executor, '--executor-instance', executor_instance
+            '--src-revs', f"{in_dataset_id}@{his_task_id}", '--asset-cache-dir', asset_cache_dir, '--config-file',
+            config_file, '--executor', executor, '--executor-instance', executor_instance
```
需要改为task-config-file吗

fenrir-z replied:
感觉没有必要。
在 command 的语境里面，config file 的意义是相当明确的。

phoenix-xhuang replied:
专门查了下，这里的config_file是从gen_executor_config_path生成的，里面的文件名就叫task_config.yaml，看上去加上task前缀更可读。
另外还有个问题：以前只executor一个参数，现在又加上了executor-instance，这两个之间也是容易混淆，考虑改名下吗？比如executor->executor-image-name or executee-name；executor-instance->executant-instance-name or executant-instance

fenrir-z replied:
done

phoenix-xhuang commented on ymir/command/mir/commands/training.py:
```diff
@@ -361,17 +360,17 @@ def run_with_args(work_dir: str,
         path_binds = []
-        path_binds.append(f"-v{work_dir_in}:/in")
+        path_binds.append(f"-v{work_dir_in}:/in")  # annotations, models, train-index.tsv, val-index.tsv, config.yaml
+        path_binds.append(f"-v{asset_cache_dir}:/in/assets:ro")  # assets
```
ro是啥作用? read-only?

fenrir-z replied:
对，可以参考 docker 的官方文档关于 volume binding 的部分

phoenix-xhuang commented on ymir/command/mir/commands/training.py:
```diff
@@ -231,6 +230,8 @@ def run_with_args(work_dir: str,
             executor_instance = f"default-training-{task_id}"
         if not tensorboard_dir:
             tensorboard_dir = os.path.join(work_dir, 'out', 'tensorboard')
+        if not asset_cache_dir:
```
这里把asset_cache_dir和executor_assets_dir混用了，应该单独创建一个变量。下面就不会出现f"-v{asset_cache_dir}:/in/assets:ro“这样看上去总是把cache绑到/in/assets的误导

fenrir-z replied:
done

Commit:
Merge branch 'dev' into zzw/dev-cmd-train-cache

Commit:
command:
    --config-file -> --task-config-file
    --executor-instance -> --executant-name

Commit:
--executor-instance -> --executant-name
executor_instance -> executant_name

Commit:
Merge branch 'zzw/dev-cmd-train-cache' of github.com:fenrir-z/ymir into zzw/dev-cmd-train-cache

Commit:
mir training: tiny changed

Commit:
mir training: removed test set

Commit:
update

Commit:
set /in/assets and /in/models as readonly

fenrir-z committed:
Merge branch 'dev' into zzw/dev-cmd-train-cache

Commit:
Merge branch 'dev' into zzw/dev-cmd-train-cache

Commit:
update doc

Commit:
Merge branch 'zzw/dev-cmd-train-cache' of github.com:fenrir-z/ymir into zzw/dev-cmd-train-cache

under-chaos committed:
Merge branch 'dev' into zzw/dev-cmd-train-cache
